packages/@uppy/core/src/Uppy.js

```diff
@@ -178,6 +179,23 @@ class Uppy {
     return this.store.getState()
   }
 
+  patchFilesState (filesWithNewState) {
+    const existingFilesState = this.getState().files
+
+    this.setState({
+      files: {
+        ...existingFilesState,
+        ...Object.fromEntries(Object.entries(filesWithNewState).map(([fileID, newFileState]) => ([
+          fileID,
+          {
+            ...existingFilesState[fileID],
+            ...newFileState,
+          },
+        ]))),
+      },
+    })
+  }
```

Comment:
The number 1 performance issue in JS when working with many arrays/objects is using the spread operator. This function is currently _very_ memory heavy. I think we should try to rewrite this with mutation if possible.

Reply:
but yea maybe we could rewrite our state to allow using `immer`, as suggested in that article. it makes code much more readable too.

Reply:
instead of rewriting what? `setFileState`? I think that is a public facing method, so if we want to rewrite it, it would be a breaking change.

Reply:
My proposal is to keep `setState` as it is, but change these 10 spread operators we pass to `setState`. That should already be good enough? I don't like the idea of adding `patchFilesState` as it's conceptually no different from `setFileState`

Reply:
not sure what you would like to change the 10 (4?) spread operators to. You mean doing something like this instead:

```js
const newFilesState = JSON.parse(JSON.stringify(existingFilesState))

Object.entries(filesWithNewState).forEach(([fileID, newFileState]) => {
  newFileState[fileID] = { ...newFileState[fileID], ...newFileState }
})

this.setState(newFilesState)
```

Reply:
not sure if the JSONparse/stringify clone is really faster than the spread though. and it could cause unnecessary re-renders because we are also updating other files state

Reply:
I haven't used it yet but perhaps [`structuredClone`](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone) makes sense here?

Reply:
maybe, but we don't have any evidence that structuredClone is faster than JSON.stringify/parse or spread operator, and i still thimk it could cause unnecessary re-renders because we are also updating other unrelated files state

Reply:
Pretty certain it should be faster. But I don't see how things can cause unnecessary re-renders more than it currently does? Everything is being spreaded.

Reply:
If this already improves performance by a huge margin, should we not go with it? Seems a little like premature optimization if we keep trying to optimize this right away. Maybe we can leave a comment that this is the place to look if reviewing performance again in the future. If there’s a simple way to rewrite the spreads for better performance, could you share a code example?

Reply:
TBH I doubt that cloning the **whole** object tree is faster than spreading **parts** of the object tree, but I also don't have any evidence, so that would be speculation / educated guess. I agree with arturi that it doesn't make sense to prematurely optimise this, as it's already about 100x faster than the previous code and it is sufficciently fast for adding many tens of thousands of files now. I think if we start to get into the realm of adding/uploading millions of files, then we will have other bottlenecks too.